Incomplete support Array type (#2871)

Seaven opened the pull request:
The sub-element type only support INT/VARCHAR.

Main include:

* FE SQL syntax (Create table) support array type column,  like:

```
CREATE TABLE array_test (
    k1 tinyint(4) NOT NULL COMMENT "",
    k2 smallint(6) NOT NULL COMMENT "",
    k3 array<int(11)> NOT NULL COMMENT "",
    k8 double MAX NOT NULL COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(k1, k2, k3)
COMMENT "OLAP"
DISTRIBUTED BY HASH(k1) BUCKETS 5
PROPERTIES (
"storage_type" = "COLUMN",
    "replication_num" = "1"
);
```



* FE SQL syntax support ArrayLiteral, support array related function, and the array constucted function support sub-element type auto-cast, like : 

```
// sub-element type is INT
mysql> select array(1, 2, 3);
+----------------+
| array(1, 2, 3) |
+----------------+
| [1,  2,  3]   |
+----------------+
1 row in set (0.01 sec)

// the sub-element type is VARCHAR
mysql> select array(1, 2, 'a', 'xzcv');
+--------------------------+
| array(1, 2, 'a', 'xzcv') |
+--------------------------+
| ['1', '2', 'a', 'xzcv']          |
+--------------------------+
1 row in set (0.01 sec)

```

* FE update meta: Columns meta support array type, include meta serialized and deserialized, and add flags in FEMetaVersions for compatibility


* BE update result buffer, support dynamic mode for output variable length data type.
* BE support ArrayLiteral and array(*) functions, like:


```
mysql> SELECT ARRAY(k1,k3) from test.bigtable;
+-------------------+
| array(`k1`, `k3`) |
+-------------------+
| [1, 1]            |
| [3, 3]            |
| [7, 7]            |
| [6, 6]            |
| [4, 4]            |
| [2, 2]            |
| [5, 5]            |
| [8, 8]            |
+-------------------+
8 rows in set (0.02 sec)

```


And other todo:
- [ ] BE meta udpate
- [ ] BE transfer data from storage to tuple
- [ ] BE support array type related function(hash, equal.....)
- [ ] BE support array type data import
- [ ] other....(many related code, such as print value, value copy...)

Commit:
Incomplete support Array type (#2871)

The sub-element type only support INT/VARCHAR.

Main include:

* FE SQL syntax (Create table) support array type column, like 'name Array<int> not null'
* FE SQL syntax support ArrayLiteral, like 'select array(1, 2, 3, 4)'
* FE update meta: Columns meta support array type, include meta serialized and deserialized, and add flags in FEMetaVersions for compatibility
* FE support ArrayLiteral: Mainly support array related function, and the array constucted function support sub-element type auto-cast.

* BE update result buffer, support dynamic mode for output variable length data type.
* BE support ArrayLiteral and array(*) functions.

And other todo:
* BE meta udpate
* BE transfer data from storage to tuple
* BE support array type related function(hash, equal.....)
* BE support array type data import
* other....(many related code, such as print value, value copy...)

Commit:
Array type update

* support test case
* update CollectionValue null bitmap struct

Seaven committed:
Fix array type case (#2871)

imay commented on be/src/runtime/collection_value.cpp:
```diff
@@ -0,0 +1,203 @@
+
+namespace doris {
+
+int sizeof_type(const PrimitiveType &type) {
```
```suggestion
int sizeof_type(const PrimitiveType& type) {
```

imay commented on be/src/runtime/collection_value.h:
```diff
@@ -0,0 +1,123 @@
+
+class CollectionIterator;
+
+class CollectionValue {
```
comment for this class

imay commented on be/src/runtime/result_writer.cpp:
```diff
@@ -57,20 +57,15 @@ Status ResultWriter::init(RuntimeState* state) {
 
-    for (int i = 0; 0 == buf_ret && i < num_columns; ++i) {
-        void* item = _output_expr_ctxs[i]->get_value(row);
+    if (NULL == item) {
```
```suggestion
    if (item == nullptr) {
```

imay commented on be/src/udf/udf.h:
```diff
@@ -786,6 +792,31 @@ struct HllVal : public StringVal {
     void agg_merge(const HllVal &other);
 };
 
+struct CollectionVal : public AnyVal {
```
Actually, if this struct is defined, the struct can not be changed, because some UDF function will use this struct. If this is only used for array, we can make it ArrayVal?

Seaven replied:
I referenced the implementation of Impala,  nested type(array, map)  is used same struct to describe, the different is the data in struct, so I think we can also use same struct. Personally, I also think called ArrayVal and ArrayValue is good idea

imay commented on be/src/util/mysql_row_buffer.cpp:
```diff
@@ -26,6 +26,8 @@
 
 namespace doris {
 
+#define NOT_DYNAMIC_MODE  (_dynamic_mode == 0)           
```
Can you give more comments for this option to help others?

Seaven replied:
OK. Will add comment in .h

Seaven committed:
Fix Array Type code style (#2871)

morningman commented on fe/src/main/java/org/apache/doris/analysis/CreateTableStmt.java:
```diff
@@ -350,6 +353,17 @@ public void analyze(Analyzer analyzer) throws UserException {
 
+            if (columnDef.getType().isArrayType()) {
+                ArrayType tp =  (ArrayType) columnDef.getType();
+                if (!PrimitiveType.INT.equals(tp.getItemType().getPrimitiveType()) &&
```
PrimitiveType is enum class, so you can just using `==`
like:
```
tp.getItemType().getPrimitiveType() == PrimitiveType.INT
```

morningman commented on fe/src/main/java/org/apache/doris/analysis/ExpressionFunctions.java:
```diff
@@ -221,6 +221,12 @@ public LiteralExpr invoke(List<Expr> args) throws AnalysisException {
                 throw new IllegalArgumentException("Doris does't support type:" + argType);
             }
+
+            // if args all is NullLiteral
```
Why not using `argType.isNull()` and put this logic into above `if else` block?

morningman commented on fe/src/main/java/org/apache/doris/analysis/TypeDef.java:
```diff
@@ -67,6 +69,20 @@ private void analyze(Type type) throws AnalysisException {
     }
+
+    if (type.isArrayType()) {
+      if (FeMetaVersion.VERSION_CURRENT < FeMetaVersion.VERSION_74) {
```
This is not a valid way to forbid a feature...

morningman commented on fe/src/main/java/org/apache/doris/analysis/TypeDef.java:
```diff
@@ -67,6 +69,20 @@ private void analyze(Type type) throws AnalysisException {
+      }
+
+      Type itemType = ((ArrayType) type).getItemType();
+      if (itemType.isNull()) {
```
I see that you just support INT 和 VARCHAR as the item type of array? Why just checking null here?

morningman commented on fe/src/main/java/org/apache/doris/catalog/Column.java:
```diff
@@ -243,9 +272,36 @@ public TColumn toThrift() {
         tColumn.setIs_allow_null(this.isAllowNull);
         tColumn.setDefault_value(this.defaultValue);
+
+        toChildrenThrift(this, tColumn);
```
It is strange to pass `this` to a private method

morningman commented on fe/src/main/java/org/apache/doris/catalog/Column.java:
```diff
@@ -243,9 +272,36 @@ public TColumn toThrift() {
 
+    private void toChildrenThrift(Column column, TColumn tColumn) {
+        if (column.type.isArrayType()) {
+            Column children = column.getChildren().get(0);
```
Why not calling `children.toThrift()`?

morningman commented on fe/src/main/java/org/apache/doris/catalog/Column.java:
```diff
@@ -419,6 +485,14 @@ public void write(DataOutput out) throws IOException {
 
         Text.writeString(out, comment);
+
+        if (Catalog.getCurrentCatalogJournalVersion() >= FeMetaVersion.VERSION_74) {
```
Since Column is already support GSON serde, we can change the write method to GSON.
And write method does not need to judge the FeMetaVersion

Seaven replied:
OK, we can refactor the serialized and deserialized method of Column in other issuses?

morningman commented on fe/src/main/java/org/apache/doris/catalog/PrimitiveType.java:
```diff
@@ -56,6 +56,7 @@
     TIME("TIME", 8, TPrimitiveType.TIME),
     // we use OBJECT type represent BITMAP type in Backend
     BITMAP("BITMAP", 16, TPrimitiveType.OBJECT),
+    ARRAY("ARRAY", 16, TPrimitiveType.ARRAY),
```
```suggestion
    ARRAY("ARRAY", 17, TPrimitiveType.ARRAY),
```

Seaven replied:
The slot_size is used for Tuple to compute memory layout, why need change to 17?

Seaven replied:
The slot_size is used for Tuple to compute memory layout, why need change to 17?

morningman replied:
Oh, I misunderstood..

morningman commented on fe/src/main/java/org/apache/doris/catalog/ScalarType.java:
```diff
@@ -141,6 +141,9 @@ public static ScalarType createType(PrimitiveType type) {
             case LARGEINT:
                 return LARGEINT;
+            case ARRAY:
+                // can't static, because the same array may cause circular dependency
```
What does this mean?

Seaven replied:
Nested type need decide sub-element type in runtime, return a static object will take repeatedly set the sub-element type of same object in many places......

Seaven replied:
Nested type need decide  and set sub-element type in runtime, return a static object will take repeatedly set the sub-element type of same object in many places......

Seaven replied:
Nested type need decide sub-element type in runtime, return a static object will take repeatedly set the sub-element type of same object in many places......

morningman commented on fe/src/main/java/org/apache/doris/catalog/Type.java:
```diff
@@ -43,11 +43,8 @@
-    // we exceeded the stack space in the scanner (which uses recursion for dealing with
-    // nested types) at a nesting depth between 200 and 300 (200 worked, 300 crashed).
-    public static int MAX_NESTING_DEPTH = 100;
+    // Maximum nesting depth of a type.
```
Why we just support 2?

Seaven replied:
Just limit the depth of nested types, deeper nested type require a lot of extra development, and I don't think we need to support more than 2 depths, like array<map<array<int>>>......

morningman commented on fe/src/main/java/org/apache/doris/rewrite/FEFunctions.java:
```diff
@@ -421,4 +422,21 @@ public static LiteralExpr ifNullBigInt(LiteralExpr first, LiteralExpr second) th
+    }
+
+    @FEFunction(name = "array", argTypes = {"VARCHAR"}, returnType = "ARRAY")
+    public static LiteralExpr array(StringLiteral... exprs) throws AnalysisException {
```
better to call arrayString()?

Seaven replied:
Em...the method is same with array(IntLiteral.....) in fact, maybe will same with array(BigintLiteral.....)、array(BoolLiteral.....) in future......I think better way is refactor the annotation for support a list of types

Seaven commented on fe/src/main/java/org/apache/doris/common/FeMetaVersion.java:
```diff
@@ -157,6 +157,8 @@
     // broker persist isAlive
     public static final int VERSION_73 = 73;
+    // array type
+    public static final int VERSION_74 = 74;
```
But array type support is incomplete(lack of storage), change the VERSION_CURRENT will take some problems, special for meta compatibility

Seaven commented on fe/src/main/cup/sql_parser.cup:
```diff
@@ -3697,6 +3700,17 @@ function_call_expr ::=
+  :}
+  | KW_ARRAY LPAREN RPAREN
+  {:
+    RESULT = new ArrayLiteral();
```
For create a array without any sub-element

Seaven commented on be/src/runtime/raw_value.cpp:
```diff
@@ -300,7 +314,34 @@ void RawValue::write(const void* value, void* dst, const TypeDescriptor& type, M
+                if (!src_iter.is_null()) {
+                    // write children 
+                    write(src_iter.value(), val_iter.value(), children_type, pool);    
+                }
```
Need,  has resolve it through copy null_signs in line 331

Seaven commented on be/src/util/mysql_row_buffer.cpp:
```diff
@@ -109,15 +132,24 @@ int MysqlRowBuffer::push_tinyint(int8_t data) {
 
-    int length = snprintf(_pos + 1, MAX_TINYINT_WIDTH + 2, "%d", data);
+    int size_bytes = 0;
+    if (NOT_DYNAMIC_MODE) {
```
Good idea, will modify it.

Seaven commented on be/src/runtime/collection_value.cpp:
```diff
@@ -0,0 +1,204 @@
+
+int sizeof_type(const PrimitiveType& type) {
+    switch (type) {
+        case TYPE_INT:
```
Just support INT/String in current version, but FE will auto cast to int if sub-element type is   tiny/small/bigint,  maybe will support other types in future.

Seaven committed:
Fix Array Type code bugs (#2871)

kangpinghuang commented on be/src/common/object_pool.h:
```diff
@@ -18,6 +18,7 @@
 #ifndef DORIS_BE_SRC_COMMON_COMMON_OBJECT_POOL_H
 #define DORIS_BE_SRC_COMMON_COMMON_OBJECT_POOL_H
 
+#include <mutex>
```
useless?

Seaven replied:
use for std::lock_guard in line 56

kangpinghuang commented on be/src/exprs/anyval_util.cpp:
```diff
@@ -164,10 +166,64 @@ FunctionContext::TypeDesc AnyValUtil::column_type_to_type_desc(const TypeDescrip
         break;
+    case TYPE_ARRAY: {
+        out.type = FunctionContext::TYPE_ARRAY;
+        for (auto&& t : type.children) {
```
why use right value reference here?

Seaven replied:
Em，the better way is `const auto &`, will modify it.

kangpinghuang commented on be/src/exprs/anyval_util.h:
```diff
@@ -296,6 +302,8 @@ class AnyValUtil {
 
     static FunctionContext::TypeDesc column_type_to_type_desc(const TypeDescriptor& type);
 
+    static PrimitiveType function_type_to_primitive_type(const FunctionContext::Type& type);
```
what is this api for?

Seaven replied:
useless function, will delete it.

kangpinghuang commented on be/src/util/mysql_row_buffer.cpp:
```diff
@@ -261,21 +321,28 @@ int MysqlRowBuffer::push_string(const char* str, int length) {
     }
 
-    int1store(_pos, 251);
+    int1store(_pos, 251);    
```
```suggestion
    int1store(_pos, 251);
```

kangpinghuang commented on be/src/util/mysql_row_buffer.h:
```diff
@@ -57,13 +57,44 @@ class MysqlRowBuffer {
+     *     mrb.push_bigint(900000);
+     * 
+     * In normal mode, the buffer contains three column:
+     *  1-'5'-3-'120'-6-'-30000' 
```
Any documents for this example?
I do not quite understand.

Seaven replied:
Em......Mysql protocol like this:

```
| flag | (size) | value | flag | (size) | value | ......

The flag means value's length or null value:
If value is NULL, flag is 251
If value's length < 250, flag is the value's length
If 250 < value's length < 65536, flag is 252 and the next two byte is length
If 65536 < value's length < 16777216 , flag is 253 and the next three byte is length
If 16777216 < value's length, flag is 254 and the next eighth byte is length
```

I will update the comment.

Seaven replied:
Em......Mysql protocol like this:

```
| flag | (length) | value | flag | (length) | value | ......
^-------A column-------^

The flag means value's length or null value:
If value is NULL, flag is 251
If value's length < 250, flag is the value's length
If 250 < value's length < 65536, flag is 252 and the next two byte is length
If 65536 < value's length < 16777216 , flag is 253 and the next three byte is length
If 16777216 < value's length, flag is 254 and the next eighth byte is length
```

I will update the comment.

Seaven committed:
Array type fix code style and update comments (#2871)

Seaven committed:
merge code and reset collection_val to array_val (#2871)

Seaven committed:
Fix sub-query failed case (#2871)

Seaven committed:
Fix sub-query failed local case (#2871)

ArrayType will cause some sub-query to failed. The reason is that
the ArrayType will used in subquery, and the ArrayType mean mutli-rows
but not a scalar-value in subquery's context.
